command_line/anvil_correction.py

```diff
@@ -0,0 +1,325 @@
+# coding: utf-8
+
+"""
+Correct integrated intensities to account for attenuation by a diamond anvil cell.
+
+High pressure X-ray diffraction experiments often involve a diamond anvil pressure
+cell, in which the sample is sandwiched between two anvils, effectively parallel flat
+plates of diamond.  The passage of the incident and diffracted beam through the
+anvils results in attenuation of both beams by the diamond by an amount that is
+dependent on the path length of each beam through each anvil.
+
+This utility calculates these path lengths and boosts the integrated reflection
+intensities to remove the calculated effect of the diamond attenuation.
+
+It is intended that this program be used to correct reflection intensities after
+integration but before scaling.  Call it on the output of dials.integrate.
+
+Examples::
+
+  dials.anvil_correction integrated.expt integrated.refl
+
+  dials.anvil_correction integrated.expt integrated.refl thickness=1.2 normal=1,0,0
+"""
+
+from __future__ import absolute_import, division, print_function
+
+import logging
+import os
+import sys
+import warnings
+
+import numpy as np
+from scipy.spatial.transform import Rotation
+
+from cctbx.eltbx import attenuation_coefficient
+from dials.array_family import flex
+import dials.util.log
+from dials.util.multi_dataset_handling import (
+    parse_multiple_datasets,
+    sort_tables_to_experiments_order,
+)
+from dials.util.options import flatten_experiments, flatten_reflections, OptionParser
+from dxtbx.model import Experiment
+import libtbx.phil
+
+try:
+    from typing import List, Sequence, SupportsFloat
+except ImportError:
+    pass
+else:
+    Vector = Sequence[SupportsFloat]
+
+
+filename_root = os.path.splitext(os.path.basename(__file__))[0]
+logger = logging.getLogger("dials.%s" % filename_root)
+
+phil_scope = libtbx.phil.parse(
+    u"""
+    anvil
+        .caption = "Properties of the mounted diamond anvils"
+    {
+        density = 3510
+            .type = float
+            .help = '''The density of the anvil material in kg per cubic metre.
+            The default is the typical density of synthetic diamond.'''
+
+        thickness = 1.5925
+            .type = float
+            .help = '''The thickness in mm of each anvil in the pressure cell.
+            The default is the thickness of the pressure cells in use on
+            beamline I19 at Diamond Light Source.'''
+
+        normal = 0, 1, 0
+            .type = floats(size=3)
+            .help = '''A 3-vector representing the normal to the anvil surfaces in the
+            laboratory frame when the goniometer is at zero datum, i.e. the
+            axes are all at zero degrees.  The vector may be given
+            un-normalised.'''
+    }
+
+    output {
+        experiments = None
+            .type = path
+            .optional = True
```

Comment:
is this the same `optional` thing we discussed elsewhere? intuitively correct, but really inappropriate?

Reply:
Sorry, I'm belatedly picking up on this.  I don't remember this discussion...  What's the problem?

Reply:
https://github.com/dials/dials/pull/1157#discussion_r380094969

Reply:
I.e. `.optional` has absolutely no effect here, so it would be better to remove it to avoid potential confusion.

Reply:
Ah yes, thanks for the reminder.